### PR TITLE
nixos/postgresql: add ensureRoles option to ensureUsers

### LINKED ISSUE #204189: postgresql: allow granting roles in `ensurePermissions`
## Issue description
Currently, the postgresql module allows ensuring the presence of users through `services.postgresql.ensureUsers.*` and making sure a user has certain permissions through `*.ensurePermissions`.

However, this only supports granting permissions in the pattern `GRANT ${perm} ON ${obj} TO ${user}`.

Postgres also allows granting roles to users through the syntax `GRANT ${role} TO ${user}`.

This is especially useful when using the fairly recent [builtin roles](https://www.postgresql.org/docs/15/predefined-roles.html), which can for example be used to give a user for monitoring appropriate limited permissions.

### Implementation ideas
The cleanest way I can think of doing this is adding `services.postgresql.ensureUsers.*.ensureRoles` as a list of strings. All role names specified will then be granted in a similar way as already happens for the permissions.

I'd be up for trying to submit a PR for this change myself, but it'd be my first time contributing to nix, so I'm happy about feedback.

### PR DESCRIPTION
**Pull request by vs49688**
## Description of changes

Adds an `ensureRoles` option to `services.postgresql.ensureUsers`. Mostly useful to for granting the new predefined roles in Postgres 14.

Use like so:

```nix
    services.postgresql.ensureUsers = [
      { name = "backup"; ensureRoles = [ "pg_read_all_data" ]; }
    ];
```

Fixes #204189.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
- Aren't users and roles the same thing in postgres?
- Is this a declarative relation, or an imperative operation that copies permissions? In the latter case, we can't make a declarative option for it.
- It should remove declaratively specified roles when they are removed from the configuration. This implies extra state, that would best be tracked on the NixOS host or in the database. See also the discussion at...
- ... possible duplicate: https://github.com/NixOS/nixpkgs/pull/203474

**Comment by vs49688** [comment]
> * Aren't users and roles the same thing in postgres?

Yes, I believe the `USER` syntax is only provided for compatibility. I didn't want to change the existing module too heavily.

> * Is this a declarative relation, or an imperative operation that copies permissions? In the latter case, we can't make a declarative option for it.

Kind of? No permissions are copied per se, it's just effectively saying "this user/role inherits everything of this other role". 

> * It should remove declaratively specified roles when they are removed from the configuration. This implies extra state, that would best be tracked on the NixOS host or in the database. See also the discussion at...

It's the same deal for `ensureUsers` - users aren't deleted from the DB when they're removed from the NixOS config.
I'd personally be a bit apprehensive about potentially deleting database state without user interaction.

> * ... possible duplicate:

**Comment by roberth** [comment]
> I'd personally be a bit apprehensive about potentially deleting database state without user interaction.

For an _ensure_ option, I wholeheartedly agree, but the issue has been [raised](https://github.com/NixOS/nixpkgs/issues/206467#issuecomment-1355889925) that ensure-style options cause stale state, so we may do well to avoid ensure-style options in the first place, and go for a solution where the declared state is managed entirely, consistently by NixOS.

Whether _ensure_ is good enough may be subjective, so I don't think it's a deal breaker.

Could you add this to the test suite?
I think the docs can also be improved. The role/user non-distinction should be addressed, and I think something went wrong with the explanation. Also it'd be nice to explain the effect in addition to the syntax produced.

**Comment by vs49688** [comment]
> > I'd personally be a bit apprehensive about potentially deleting database state without user interaction.
> 
> For an _ensure_ option, I wholeheartedly agree, but the issue has been [raised](https://github.com/NixOS/nixpkgs/issues/206467#issuecomment-1355889925) that ensure-style options cause stale state, so we may do well to avoid ensure-style options in the first place, and go for a solution where the declared state is managed entirely, consistently by NixOS.
> 
> Whether _ensure_ is good enough may be subjective, so I don't think it's a deal breaker.
> 

I agree it's a bit iffy and does cause stale state under some circumstances - but what's the alternative? Short of doing some hackage such as with dumping all the existing roles/users at re-build time and comparing them to what's there (which, in turn has its own state issues), I don't really see a nicer way.

> Could you add this to the test suite? I think the docs can also be improved.

Can do.

> The role/user non-distinction should be addressed, and I think something went wrong with the explanation. Also it'd be nice to explain the effect in addition to the syntax produced.

What if I rename `ensureUsers` to `ensureRoles`, change this PR's `ensureRoles` to `memberOf` or something similar?

**Comment by vs49688** [comment]
> > For an _ensure_ option, I wholeheartedly agree, but the issue has been [raised](https://github.com/NixOS/nixpkgs/issues/206467#issuecomment-1355889925) that ensure-style options cause stale state, so we may do well to avoid ensure-style options in the first place, and go for a solution where the declared state is managed entirely, consistently by NixOS.
> > Whether _ensure_ is good enough may be subjective, so I don't think it's a deal breaker.
> 
> I agree it's a bit iffy and does cause stale state under some circumstances - but what's the alternative? Short of doing some hackage such as with dumping all the existing roles/users at re-build time and comparing them to what's there (which, in turn has its own state issues), I don't really see a nicer way.
> 

I just had a thought on this - we (NixOS) explicitly create a database (e.g. `nixos`) and store current state there. Then at activation-time, we simply compare against that. That's probably outside the scope of this PR though.

**Comment by roberth** [comment]
> What if I rename `ensureUsers` to `ensureRoles`

We should stick to upstream terminology, so I'd be in favor of such a change.

> and change the new `ensureRoles` to `memberOf` or something similar?

When I scanned the docs for this feature, there wasn't a term that jumped out, to name this relation, but I'm sure they've named it somewhere. Ideally we use a name they already use, even if the name isn't documented (and not terrible, because then they might change it or document it differently later). Maybe it shows in postgres' internal schema; something like `pg_roles` or whatever the table is named.

> we (NixOS) explicitly create a database (e.g. `nixos`) and store current state there. Then at activation-time, we simply compare against that. That's probably outside the scope of this PR though.

Yeah, I think it should look like that, although we could possibly improve it by taking it in either of two directions:
 - Use a file so that we can work towards a reusable framework for managing the state that is needed for declarative removals - reusable outside the context of postgres. We'd have infrastructure for doing similar things with any other kind of service, including things like NixOS mutable users (already implemented correctly - probably not worth refactoring, but I'm sure we can still apply the framework to other areas that aren't database servers).
 - Turn this into a postgres admin tool that is reusable outside the context of NixOS, and can be applied to any database; not just a local one. This also helps NixOS, for example when users need to scale their service and move the database out of the app server into a different server or managed db. Ideally this tool already exists and we just translate/automate it in NixOS.

**Comment by vs49688** [comment]
> > What if I rename `ensureUsers` to `ensureRoles`
> 
> We should stick to upstream terminology, so I'd be in favor of such a change.
> 

:+1: 

> > and change the new `ensureRoles` to `memberOf` or something similar?
> 
> When I scanned the docs for this feature, there wasn't a term that jumped out, to name this relation, but I'm sure they've named it somewhere. Ideally we use a name they already use, even if the name isn't documented (and not terrible, because then they might change it or document it differently later). Maybe it shows in postgres' internal schema; something like `pg_roles` or whatever the table is named.
> 

Postgres uses `pg_group` internally (irrelevant rows removed):

```
postgres=> SELECT * FROM pg_catalog.pg_roles;
          rolname          | rolsuper | rolinherit | rolcreaterole | rolcreatedb | rolcanlogin | rolreplication | rolconnlimit | rolpassword | rolvaliduntil | rolbypassrls | rolconfig |  oid
---------------------------+----------+------------+---------------+-------------+-------------+----------------+--------------+-------------+---------------+--------------+-----------+--------
 pg_read_all_data          | f        | t          | f             | f           | f           | f              |           -1 | ********    |               | f            |           |   6181
 backup                    | f        | t          | f             | f           | t           | f              |           -1 | ********    |               | f            |           | 162320

postgres=> SELECT * FROM pg_catalog.pg_group;
          groname          | grosysid | grolist
---------------------------+----------+----------
 pg_read_all_data          |     6181 | {162320}

postgres=> SELECT * FROM pg_catalog.pg_roles WHERE oid=162320;
 rolname | rolsuper | rolinherit | rolcreaterole | rolcreatedb | rolcanlogin | rolreplication | rolconnlimit | rolpassword | rolvaliduntil | rolbypassrls | rolconfig |  oid
---------+----------+------------+---------------+-------------+-------------+----------------+--------------+-------------+---------------+--------------+-----------+--------
 backup  | f        | t          | f             | f           | t           | f              |           -1 | ********    |               | f            |           | 162320
```

Looking at https://www.postgresql.org/docs/current/role-membership.html, there's "role membership", "group role", and "member roles". Some options:

```nix
    services.postgresql.ensureRoles = [ # Changed from ensureUsers
      { name = "backup"; ensureRoles = [ "pg_read_all_data" ]; }
      { name = "backup"; memberOf = [ "pg_read_all_data" ]; }
      { name = "backup"; memberships = [ "pg_read_all_data" ]; }
      { name = "backup"; groups = [ "pg_read_all_data" ]; }
      { name = "backup"; additionalRoles = [ "pg_read_all_data" ]; }
    ];
```

> > we (NixOS) explicitly create a database (e.g. `nixos`) and store current state there. Then at activation-time, we simply compare against that. That's probably outside the scope of this PR though.
> 
> Yeah, I think it should look like that, although we could possibly improve it by taking it in either of two directions:
> 
>     * Use a file so that we can work towards a reusable framework for managing the state that is needed for declarative removals - reusable outside the context of postgres. We'd have infrastructure for doing similar things with any other kind of service, including things like NixOS mutable users (already implemented correctly - probably not worth refactoring, but I'm sure we can still apply the framework to other areas that aren't database servers).
> 
>     * Turn this into a postgres admin tool that is reusable outside the context of NixOS, and can be applied to any database; not just a local one. This also helps NixOS, for example when users need to scale their service and move the database out of the app server into a different server or managed db. Ideally this tool already exists and we just translate/automate it in NixOS.

All definitely outside the scope of this PR, but both options are worth looking into.

**Comment by roberth** [comment]
> Some options:

Broadly the choice seems to be between just "role", "group" and "member".
- role is not very descriptive
- group seems to suggest an entity more than a relation between entities. It could be defined as a set of memberships, but it represents the inverse relation. I think we want to declare at the member which groups it is part of rather than declare at the group which members it has.
- memberOf or membership don't suffer from this, so I think they're most appropriate. Maybe of the two, `memberships` is preferable because it's plural and the value is a list.


I think it'd be good to highlight that we're still only ever adding to the memberships, by repeating `ensure`. How about `ensureMemberships`?
Just `memberships` may suggest that we are _ensuring_ that a role exists, but with an _exact_ set of memberships.
I *think* most `ensureUsers.<name>.<attr>` `attr`s intentionally have an `ensure` prefix for this reason, although I doubt that `ensureClauses` needs it.

**Comment by vs49688** [comment]
I just started attacking this - turns out the is a difference between a user and a role. Users have the `LOGIN` permission applied by default. Do we still want `ensureUsers` renamed to `ensureRoles`?

File: nixos/modules/services/databases/postgresql.nix [hunk 1]
```diff
@@ -565,7 +583,11 @@ in
               )
               cfg.ensureUsers
             }
-          '';
+          '' + concatMapStrings (user: let
+            userRoles = concatStringsSep "," user.ensureRoles;
+          in ''
+            $PSQL -tAc 'GRANT "${userRoles}" TO "${user.name}"'
+          '') (filter (user: length user.ensureRoles != 0) cfg.ensureUsers);
```
Comment (Ma27) [inline nixos/modules/services/databases/postgresql.nix:590]: AFAIU GRANTs aren't dropped here when removed from the configuration, right?
I'm not sure if there's a reasonable way to implement that, but if not then it should be warned about this behavior in at least the option's description.

**Review by Ma27 — COMMENTED** [review]
A testcase would also be nice.

File: nixos/modules/services/databases/postgresql.nix (hunk 1 above)
Comment (vs49688) [reply nixos/modules/services/databases/postgresql.nix:590]: > AFAIU GRANTs aren't dropped here when removed from the configuration, right?

Correct.

> I'm not sure if there's a reasonable way to implement that, but if not then it should be warned about this behavior in at least the option's description.

There's not, we're dealing with an inherently stateful system. Nix can only do so much in that regard.

I had planned to add it to the description, I just wanted to nail down what we actually want with this PR before I continued.

**Comment by vs49688** [comment]
> A testcase would also be nice.

Indeed, I planned to add one once I'd finished the rest.

> I just started attacking this - turns out there is a difference between a user and a role. Users have the `LOGIN` permission applied by default. Do we still want `ensureUsers` renamed to `ensureRoles`?

Btw, do you have an opinion/preference on this? I never received a response, which is why I stopped working on this PR.

**Comment by RaitoBezarius** [comment]
> > I just started attacking this - turns out there is a difference between a user and a role. Users have the `LOGIN` permission applied by default. Do we still want `ensureUsers` renamed to `ensureRoles`?
> 
> Btw, do you have an opinion/preference on this? I never received a response, which is why I stopped working on this PR.

I think I prefer `ensureRoles`, DBA PGSQL people knows what it means and how it differs, also it's synced with upstream docs.

**Comment by Ma27** [comment]
I agree with @roberth's suggestion about a state management solution in the longterm, however
* `ensureUsers` is broken for quite a bunch of services apparently on master
* a new NixOS release is knocking on the door

so we'll need a solution.

Unless we go for a full revert this is IMHO the most reasonable solution[1] as it doesn't expand on the problematic `ensure*` approach.

> I think I prefer ensureUsers and ensureRoles separated, DBA PGSQL people knows what it means and how it differs, also it's synced with upstream docs.

Just to be clear, you mean the way it's currently implemented in this PR, correct? If yes, I agree.

Inventing our own names means that everyone has to map the NixOS names onto the Postgres names when reading upstream docs and that seems like a very bad idea. And `roles` is the correct term for what we want to do now. Finally, I wouldn't want to bikeshed `ensureUsers` and block this PR on that now (and in the future the time is probably better invested into alternatives).


[1] Though an even simpler solution would be to unconditionally add the CREATE permission to restore the pre-15 behavior.

**Comment by RaitoBezarius** [comment]
Yes, I agree with you Ma27.

Le lun. 23 oct. 2023, 22:02, Maximilian Bosch ***@***.***> a
écrit :

**Comment by vs49688** [comment]
Okay I've had a look at some of the issues on master, and now I'm a bit confused (granted it is very late, so I'm looking at this through sleepy eyes).

PG15 doesn't add the `CREATE` permission on the `public` by default schema, right? That's not a role, so how would this help?

This PR adds `ensureUsers.*.ensureRoles = ["role1" "role2"]`, which expands to `GRANT role1,role2 TO user;`

(Unless you're abusing the syntax with `ensureRoles = ["CREATE"]` expanding to `GRANT CREATE TO ${user}`)?

**Comment by RaitoBezarius** [comment]
> Okay I've had a look at some of the issues on master, and now I'm a bit confused (granted it is very late, so I'm looking at this through sleepy eyes).
> 
> PG15 doesn't add the `CREATE` permission on the `public` by default schema, right? That's not a role, so how would this help?
> 
> This PR adds `ensureUsers.*.ensureRoles = ["role1" "role2"]`, which expands to `GRANT role1,role2 TO user;`
> 
> (Unless you're abusing the syntax with `ensureRoles = ["CREATE"]` expanding to `GRANT CREATE TO ${user}`)?

Excellent point. I am travelling ATM, I don't have PGSQL docs but this only hints to me that ensureRoles can do more and should have a more general name?

**Comment by Ma27** [comment]
> (Unless you're abusing the syntax with ensureRoles = ["CREATE"] expanding to GRANT CREATE TO ${user})?

This is a good point. I guess I was too focused on the user vs role thing that I missed we have a third entity that's relevant here, permissions :facepalm: That's kinda embarassing and I'm very sorry for that.

Also unless I'm missing something that wouldn't even be sufficient, you'd have to commit crimes like `ensureRoles = [ "CREATE ON  SCHEMA public"];` (caution: untested!) 

This means that

* I'm against the idea of this PR. The sole reason why I was OK with adding another `ensure*`-option is to fix an immediate issue.
* I see effectively two choices (short of dropping these options alltogether):
  * Make it possible to declare db ownership in `ensureUsers` and force everybody to use that.
  * Something like

```diff
diff --git a/nixos/modules/services/databases/postgresql.nix b/nixos/modules/services/databases/postgresql.nix
index 2d4ef0563182..f3fbabfab5a0 100644
--- a/nixos/modules/services/databases/postgresql.nix
+++ b/nixos/modules/services/databases/postgresql.nix
@@ -559,7 +559,10 @@ in
                 let
                   userPermissions = concatStringsSep "\n"
                     (mapAttrsToList
-                      (database: permission: ''$PSQL -tAc 'GRANT ${permission} ON ${database} TO "${user.name}"' '')
+                      (database: permission: ''
+                        $PSQL -tAc 'GRANT ${permission} ON ${database} TO "${user.name}"'
+                        $PSQL -U postgres -d wiki -tAc 'GRANT CREATE ON SCHEMA public TO "${user.name}"'
+                      '')
                       user.ensurePermissions
                     ); 
```

  though we'd need a way to determine the database here properly (I hardcoded it to fix the `wiki-js` test which I just used as an example).

**Comment by vs49688** [comment]
Yeah, then I'm not sure what the immediate path forward is here. Unfortunately, I have neither the time nor the patience to come up with a proper solution, getting it reviewed (potentially months), dealing with the inevitable bikeshedding, etc. 

I'm using this as a standalone module as it solves an actual problem I had (a backup user needing access to everything).
If this is not the right way forward, that's fine. I'll close this PR now.

Standalone module, for reference:
```nix
{ config, lib, ... }: {
  options.services.postgresql.ensureUsers = with lib; mkOption {
    type = types.listOf (types.submodule {
      options = {
        ensureRoles = mkOption {
          type = types.listOf types.str;
          default = [];
          description = lib.mdDoc ''
            Roles to ensure for the user, specified as a list of strings.

            For more information on how to specify the role, see the
            [GRANT syntax](https://www.postgresql.org/docs/current/sql-grant.html).
            The attributes are used as `GRANT ''${role}.
          '';
          example = literalExpression ''
            [
              "pg_read_all_data"
              "pg_write_all_data"
            ]
          '';
        };
      };
    });
  };

  config = {
    systemd.services.postgresql.postStart = lib.concatMapStrings (user: let
        userRoles = lib.concatStringsSep "," user.ensureRoles;
      in ''
        $PSQL -tAc 'GRANT "${userRoles}" TO "${user.name}"'
      ''
    ) (lib.filter (user: lib.length user.ensureRoles != 0) config.services.postgresql.ensureUsers);
  };
}
```

**Comment by RaitoBezarius** [comment]
> Unfortunately, I have neither the time nor the patience to come up with a proper solution, getting it reviewed (potentially months), dealing with the inevitable bikeshedding, etc.

It's unfortunate because as we are running this immediate PostgreSQL 15 problem, us, contributors, were examining your solution as the most promising one for expedited review and merge. But it's fair.

**Comment by vs49688** [comment]
> > Unfortunately, I have neither the time nor the patience to come up with a proper solution, getting it reviewed (potentially months), dealing with the inevitable bikeshedding, etc.
> 
> It's unfortunate because as we are running this immediate PostgreSQL 15 problem, us, contributors, were examining your solution as the most promising one for expedited review and merge. But it's fair.

I was probably a bit harsh, I apologise. I was just agitated.

As far as I understood, this solution can't be used to fix the pg15 problem in an elegant way, which is why I closed it.
It's about role membership, not user permissions. Unless I'm missing something?

**Pull request closed without merging**